### DISCUSSION TITLE
4 Semester 7 Woche

### DISCUSSION THREAD

**Discussion by MelonigeMelone** [MelonigeMelone/xCalendar#38] [Blogkanal]
# CI/CD Pipeline

## Aktueller Stand
Wir hatten bereits im ersten Semester eine Jenkins Pipeline eingerichtet, welche nach jedem Commit auf den Main Branch, das Projekt baut und auf einen Testserver deployed, welcher unter https://dev01.melonigemelone.de aufrufbar ist.

## Nächste Schritte
Allerdings funktioniert diese Pipeline aufgrund des Ablaufs des Githubs Token nicht mehr einwandfrei, sodass wir bereits geplant hatten, diese wieder zum aktivieren, sowie die Einbindung der Tests. Zusätzlich sollten auch bei den Pull Requetsts einen Jenkins Pipeline  überprüfen, ob das Projekt noch baut und alle Test fehlerfrei sind. Diese nächsten Schritte haben wir uns zur Aufgabe genommen diese Woche umzusetzen.

## Reaktivierung der Jenkins Pipeline (Live Demo)
Durch einen Umzug des Jenkins Server, so wie dem Ablauf des Github Tokens. Mussten wir die Pipeline noch einmal neueinrichtend im ersten Schritt, wurde die Pipeline neu angelegt, die entsprechenden Daten korrigiert und ein erfolgreicher Lauf durchgeführt.
Danach wurde über die PostBuild Action "Send build artifacts over SSH" zum entsprechenden Webserver deployed.
Der Status des Jobs wird zusätzlich in der README.md angezeigt.
Dieser Job soll nun jedes mal ausgeführt werden, wenn eine Änderung auf den Main Branch commitet wurde.
Hierzu musste ein Webhook im Github Repository erstellt werden, sodass dieses den Jenkins Job triggered.

## Test Integration
Als nächsten Schritt sollte in den Jenkins Job das Ausführen der Tests hinzugefügt werden.
Hierzu musste das Shell script um `ng test` erweitert werden, sodass vor dem Build die Tests ausgeführt werden.
Im ersten lauf sind wir hierbei auf einen Fehler gestoßen. Unsere Test werden mit Karma ausgeführt dies benötigt allerdings einen installierten Chrome Browser. Allerdings auch nach der Installation dieses kam es zu weiteren Fehler im Zusammenspiel mit dem Jenkins, welche wir gerade noch am beheben sind, sodass die Tests momentan nicht in der Job integriert sind.
Somit konnte das Monitoring der Testergebnisse auch noch nicht erfolgen.

## Jenkins Pipeline (Pull Requests)
Neben der Haupt Pipeline, welche für die Aktualisierung der Demo zuständig ist, sollen auch bei jedem Pull Requests entsprechend Jenkins Job überprüfen, ob das Projekt weiterhin erfolgreich gebaut wird und die Test (sofern integriert) einwandfrei ausgeführt werden.
Mithilfe eines Jenkins Plugin als Unterstützung konnte durch einen zweiten Webhook der nur bei Updates zu Pull Requests triggered, dieser Job getriggered werden, sodass bei jedem Update bei Pull Requests ein entsprechenden Jenkins Job ausgeführt wird. Dieser wird entsprechend im Pull Requests angezeigt. Ist der Job beendet wird der Status ebenso angezeigt nur bei erfolgreicher Beendigung des Jobs kann der Pull Requests gemerkt werden.

Damit haben wir nun eine sinnvolle CI/CD Pipeline eingerichtet. Wir werden bis zur Ende der Woche noch die Test Integration abschließen, sodass die Pipeline auch sinnvoll genutzt werden können.

**Comment by Christian-2003** [comment]
Hallo zusammen,

schade, dass die bereits aufgesetzte Pipeline aus dem letzten Semester nicht mehr funktioniert. Hoffentlich war das erneute Aktivieren nicht mit einem allzu hohen Aufwand verbunden.

Dass die Tests nicht von der Pipeline berücksichtigt werden, scheint mir fürs erste nicht allzu schlimm zu sein. Wenn ihr mit eurer anderen Pipeline das Projekt automatisch builden könnt, dann spart ihr euch erheblichen Arbeitsaufwand.

Ich freue mich auf nächste Woche. Berichtet nächstes Mal gerne, ob die Test Integration funktioniert.

**Comment by MikaK21** [comment]
Hey zusammen,

euer Blogeintrag gibt echt einen guten Überblick darüber, was ihr bisher erreicht habt und welche Probleme ihr noch lösen müsst. Es ist super zu sehen, wie ihr die Jenkins Pipeline nach dem Serverumzug und dem Ablauf des Github Tokens wieder ans Laufen bekommen habt.

Die Probleme mit Karma und dem Chrome-Browser zeigen, dass ihr euch echt reinkniet, um alles hinzubekommen. Auch die Idee, Tests zu integrieren und einen Webhook für Pull Requests einzurichten, klingt nach einem sinnvollen Schritt, um die Qualität des Projekts zu sichern.

Alles in allem seid ihr auf einem guten Weg, eine stabile und funktionierende CI/CD-Pipeline aufzubauen. Bin gespannt, wie es weitergeht und wann die Tests endlich reibungslos laufen!
